### ISSUE TITLE
Crystal takes all my RAM and CPU and continues forever

### ISSUE THREAD

**Issue by stugol** [crystal-lang/crystal#2395]
I've just installed the latest build of Crystal. When I try to compile my project, Crystal takes 90% of my RAM (>2GB) and 25%-75% of my CPU and never completes. What's going on?

If you require code to repro this, I can zip my entire project and send it to you somehow.

**Comment by asterite** [comment]
We'll need to know the crystal version you are using and your code. If possible you can try to reduce it by chopping parts of the code until you get to the smallest code that reproduces the problem.

**Comment by stugol** [comment]
I've tried that. Whenever I change the code, the problem moves to a different part. The same code works in one file and not in another. It's a [heisenbug](https://en.wikipedia.org/wiki/Heisenbug), and therefore impossible to repro.

One thing I **can** tell you: When I change code like the following
```
module I::FS
   ...
end

module I::IO
   ...
end
```
to the following
```
module I_FS
   ...
end

module I_IO
   ...
end
```
the program will once again compile! It makes no sense!

Also, if I declare the original module `I::FS` etc in my main file, there is no problem. But if I declare it in an include, it fails. Sheer madness!

I'm using the current Debian repo version of Crystal (0.15.0 [0f62400]), which would appear to be the newest version.

**Comment by ysbaddaden** [comment]
Almost the same here with Crystal 0.15.0 on Ubuntu Trusty 64bits. It took an awful lot of time to compile a project (it didn't with 0.14) and it eventually failed with #2393 :

```
/home/github/posix $ crystal build --stats --release -o bin/main src/main.cr
Parse:                             00:00:00.1608960 (  50.25MB)
Semantic (top level):              00:00:00.0910500 (  74.25MB)
Semantic (abstract def check):     00:00:00.0020230 (  74.25MB)
Semantic (type declarations):      00:00:00.0100260 (  82.25MB)
Semantic (initializers):           00:00:00.0027280 (  82.25MB)
Semantic (main):                   00:00:05.3511520 ( 572.18MB)
Semantic (cleanup):                00:00:00.0001130 ( 572.18MB)
Semantic (recursive struct check): 00:00:00.0010980 ( 580.18MB)
Codegen (crystal):                 00:00:02.1503900 ( 630.18MB)
Codegen (bc+obj):                  00:01:55.2220020 ( 630.18MB)
Codegen (linking):                /opt/crystal/src/llvm/ext/llvm_ext.o: dans la fonction « LLVMNewDIBuilder »:
llvm_ext.cc:(.text.LLVMNewDIBuilder+0x25): référence indéfinie vers « llvm::DIBuilder::DIBuilder(llvm::Module&) »
```

**Comment by ysbaddaden** [comment]
Actually it also happens with 0.14.2 (takes lots of time and linking fails). Compare with 0.13.0:

```
/home/github/posix $ crystal build --stats --release -o bin/main src/main.cr
Parse:                             00:00:00.1941020 (  50.34MB)
Semantic (top level):              00:00:00.0825340 (  74.34MB)
Semantic (abstract def check):     00:00:00.0055340 (  74.34MB)
Semantic (type declarations):      00:00:00.0020130 (  74.34MB)
Semantic (main):                   00:00:00.3971600 ( 130.40MB)
Semantic (cleanup):                00:00:00.0000070 ( 130.40MB)
Semantic (recursive struct check): 00:00:00.0007280 ( 130.40MB)
Codegen (crystal):                 00:00:00.3226970 ( 130.52MB)
Codegen (bc+obj):                  00:00:09.9701850 ( 130.52MB)
Codegen (linking):                 00:00:00.7653440 ( 130.52MB)
```

I'm investigating where the problem originates from.

**Comment by ysbaddaden** [comment]
@asterite the culprit is 9a3f239fdee1410b6002cf3a10afebaaf8648d8c (add a pass to type instance variable initializers). It hogs memory up and makes bc+obj codegen tragically slow in some cases. I'll try to reduce.

**Comment by asterite** [comment]
@ysbaddaden Thank you so much for bisecting! Is there any code (big or small) I can use to reproduce this?

**Comment by asterite** [comment]
Nevermind, just saw it happens with icr,I'll try it later on linux.

**Comment by asterite** [comment]
I wasn't able to reproduce it with icr on linux. I'll need more info.

**Comment by asterite** [comment]
@stugol Does your project depend on the compiler's source code (that is, do you do `require "compiler/crystal/**"` or something similar?)

**Comment by stugol** [comment]
No...? Why on earth would I do that?

**Comment by asterite** [comment]
Because @ysbaddaden sent me some code and it turned out that the slowness came because of some type annotations added in the compiler.

I just made [this commit](https://github.com/crystal-lang/crystal/commit/0737f765c5ab44546b7bbb76aa9e8838737046be) which improved that (his project only needs the parser, not the whole compiler).

But it seems your issue is a separate one. Please send me a zip with your project to my email address (it's on my GitHub profile) and I'll be able to help.

**Comment by stugol** [comment]
Done

**Comment by BlaXpirit** [comment]
Yes, I have also ran into this with [icr](https://github.com/greyblake/crystal-icr). Ever since 0.14 I've been unable to compile it on my VPS (ran out of memory ~1.5GB).

At least with this change it compiles, and very quickly! But, of course, that's a solution only for this specific case.

What I can reproduce, at the very least, is when compiling [icr](https://github.com/greyblake/crystal-icr) on my VPS with Crystal 0.13 everything is fine, while latest master keeps chugging for very long and then is killed due to memory overuse.

**Comment by BlaXpirit** [comment]
I can only confirm what @ysbaddaden said. Here are the final lines of my bisect session (`cr` is an alias to compiled crystal):

```
----crystal-icr----master----$ time cr build --release -o bin/icr src/icr/cli.cr
Using compiled compiler at .build/crystal
real    1m35.866s
user    1m35.337s
sys     0m0.533s
----crystal----(9a495cf...)|BISECTING----$ git bisect bad
Bisecting: 2 revisions left to test after this (roughly 2 steps)
[c2f569c16493d1bc40633df5319e524ce6123e63] Compiler: inherit instance variable type declaration from generic modules
----crystal----(c2f569c...)|BISECTING----$ make
CRYSTAL_CONFIG_PATH=`pwd`/src ./bin/crystal build  -o .build/crystal src/compiler/crystal.cr
Using compiled compiler at .build/crystal
----crystal-icr----master----$ time cr build --release -o bin/icr src/icr/cli.cr
Using compiled compiler at .build/crystal
real    0m8.147s
user    0m8.097s
sys     0m0.160s
----crystal----(c2f569c...)|BISECTING----$ git bisect good
```
> There are only 'skip'ped commits left to test.
> The first bad commit could be any of:
> 92b758f65fd1c53f32d1bf8fb4eedc027f268fae
> 9a3f239fdee1410b6002cf3a10afebaaf8648d8c
> d3b6f33236e77e75f74dde2ec2f0e140477766c1
> f018c1f1644c45144a8c5dd44a97ee543e72a9f0
> 9a495cf2bf4dc9f81bea9808a3b0c33de7d0d3f0
> We cannot bisect more!

**Comment by asterite** [comment]
Reduced:

```crystal
# can be run with --prelude=empty
struct Nil
  def method(x : Problem)
    0
  end
end

class Foo
  macro def method(x : Problem) : Int32
    {% for ivar in @type.instance_vars %}
      @{{ivar.id}}.method(x)
    {% end %}
    42
  end
end

class Problem
end

module Moo
  class Problem
  end

  class Bar < Foo
    @foo : Foo?
  end
end

Moo::Bar.new.method(Problem.new)
```

The current way we implement macro defs involves copying the `def` into subclasses. The problem is that method argument types were being looked up in the subclasses, so the restriction `x : Problem` would not match a `Problem` because it would try to match it against `Moo::Problem` in the case of `Bar` (imagine copying the `macro def` inside `Bar`).

The solution is easy, I'll push it soon.

**Comment by asterite** [comment]
Also worth noting: it didn't work in previous versions of the compiler either.

**Comment by asterite** [comment]
@BlaXpirit I just tried changing `icr` to only `require "compiler/crystal/syntax"` and indeed it compiles very fast now.

Even though it might look like a workaround, it's not. The compiler's code wasn't well organized and you had to require the whole thing, even if you just wanted to parse source. With the added type annotations, which soon will be [kind of required](https://github.com/crystal-lang/crystal/issues/2390), what happened is that the compiler was able to reach methods that it previously couldn't. For example some subclasses of `Exception` redeinfed `to_s`, and there some methods were looked up (in the compiler), and eventually the whole compiler was typed. Now I moved some classes and type annotations to the "semantic" directory, so if you don't require that the compiler won't reach it.

As I told @ysbaddaden in an email, I believe this was something good because it forced us to better organize the code (Crystal has that effect in other aspects too, by being strict with types).

I know running out of memory by trying to compile `icr` sounds crazy, but that probably happens too when trying to compiler the compiler in a machine without much memory, and that's why we want a change. The most affected project is the compiler, but I'm sure other projects will reach a similar size so it's better if we have a language that supports that :-)

In conclusion, there were two issues here, but now both should be fixed.
